src/rt-app_taskgroups.c

```diff
@@ -0,0 +1,365 @@
+#include <stdlib.h>
+#include <string.h>
+#include <mntent.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <stdio.h>
+#include <errno.h>
+
+#include "config.h"
+#include "rt-app_utils.h"
+#include "rt-app_taskgroups.h"
+
+typedef struct _taskgroup_ctrl_t
+{
+	char mount_point[FILENAME_MAX];
+} taskgroup_ctrl_t;
+
+static taskgroup_ctrl_t tg_ctrl;
+
+static int cgroup_check_cpu_controller(void)
+{
+	int dummy[3], ret = 0;
+	FILE *cgroups;
+	char buf[512];
+
+	cgroups = fopen("/proc/cgroups", "re");
+
+	if (!cgroups) {
+		perror("fopen");
+		goto err;
+	}
+
+	/* Ignore the first line as it contains the header. */
+	if (!fgets(buf, sizeof(buf), cgroups)) {
+		perror("fgets");
+		goto err;
+	}
+
+	while (!feof(cgroups)) {
+		if (fscanf(cgroups, "%s %d %d %d", buf, &dummy[0], &dummy[1],
+			   &dummy[2]) < 4) {
+			perror("fscanf");
+			goto err;
+		}
+
+		if (!strcmp(buf, "cpu"))
+			goto done;
+	}
+err:
+	ret = -1;
+done:
+	if (cgroups)
+		fclose(cgroups);
+
+	return ret;
+}
+
+static int cgroup_get_cpu_controller_mount_point(void)
+{
+	struct mntent *ent;
+	FILE *mounts;
+	int ret = -1;
+
+	mounts = fopen("/proc/mounts", "re");
+
+	if (!mounts) {
+		perror("fopen");
+		return ret;
+	}
+
+	while (ent = getmntent(mounts)) {
+		if (strcmp(ent->mnt_type, "cgroup"))
+			continue;
+
+		if (!hasmntopt(ent, "cpu"))
+			continue;
+
+		strcpy(tg_ctrl.mount_point, ent->mnt_dir);
+
+		log_debug("cgroup cpu controller mountpoint [%s] found", ent->mnt_dir);
+		ret = 0;
+		break;
+	}
+
+	fclose(mounts);
+	return ret;
+}
+
+static int cgroup_attach_task(char *name)
+{
+	char *path;
+	FILE *tasks;
+	int ret = -1;
+
+	path = malloc(FILENAME_MAX);
+	if (!path) {
+		perror("malloc");
+		goto error;
+	}
+
+	sprintf(path, "%s", tg_ctrl.mount_point);
+
+	if (!name)
+		goto next;
+
+	if (strlen(name) > FILENAME_MAX - strlen(path) - 1) {
+		log_error("cgroup [%s] no space to create cgroup filename", name);
+		goto error;
+	}
+
+	strcat(path, name);
+next:
+	if (strlen("/tasks") > FILENAME_MAX - strlen(path) - 1) {
+		log_error("cgroup [%s] no space to create /tasks filename", name);
+		goto error;
+	}
+
+	strcat(path, "/tasks");
```

Comment:
Ditto.
Also, can't you merge the two above?

Reply:
Yes I can! Earlier versions called cgroup_attach_task() with NULL for the root taskgroup but this has been changed to "/" in the meantime.

So this would work:
diff --git a/src/rt-app_taskgroups.c b/src/rt-app_taskgroups.c
index 82cfa1663370..2bcfee1bf293 100644
--- a/src/rt-app_taskgroups.c
+++ b/src/rt-app_taskgroups.c
@@ -88,7 +88,7 @@ static int cgroup_get_cpu_controller_mount_point(void)
 
 static int cgroup_attach_task(char *name)
 {
-       char *path;
+       char *path, *file;
        FILE *tasks;
        int ret = -1;
 
@@ -98,24 +98,14 @@ static int cgroup_attach_task(char *name)
                goto error;
        }
 
-       sprintf(path, "%s", tg_ctrl.mount_point);
+       file = strcmp(name, "/") ? "/tasks" : "tasks";
 
-       if (!name)
-               goto next;
-
-       if (strlen(name) > FILENAME_MAX - strlen(path) - 1) {
+       if (strlen(tg_ctrl.mount_point) + strlen(name) + strlen(file) > FILENAME_MAX - 1) {
                log_error("cgroup [%s] no space to create cgroup filename", name);
                goto error;
        }
 
-       strcat(path, name);
-next:
-       if (strlen("/tasks") > FILENAME_MAX - strlen(path) - 1) {
-               log_error("cgroup [%s] no space to create /tasks filename", name);
-               goto error;
-       }
-
-       strcat(path, "/tasks");
+       sprintf(path, "%s%s%s", tg_ctrl.mount_point, name, file);
 
        tasks = fopen(path, "we");
        if (!tasks) {

Reply:
Cool.